src/libraries/Common/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Pkcs12.cs

```diff
@@ -0,0 +1,909 @@
+                        };
+                    }
+
+                    bagState.ConfirmPassword();
+                }
+
+                AsnValueReader outer = new AsnValueReader(authSafeContents, AsnEncodingRules.BER);
+                AsnValueReader reader = outer.ReadSequence();
+                outer.ThrowIfNotEmpty();
+
+                ReadOnlyMemory<byte> rebind = pfxAsn.AuthSafe.Content;
+                bagState.Init(loaderLimits);
+
+                int? workRemaining = loaderLimits.TotalKdfIterationLimit;
+
+                while (reader.HasData)
+                {
+                    ContentInfoAsn.Decode(ref reader, rebind, out ContentInfoAsn safeContentsAsn);
+
+                    ReadOnlyMemory<byte> contentData;
+
+                    if (safeContentsAsn.ContentType == Oids.Pkcs7Data)
+                    {
+                        contentData = Helpers.DecodeOctetStringAsMemory(safeContentsAsn.Content);
+                    }
+                    else if (safeContentsAsn.ContentType == Oids.Pkcs7Encrypted)
+                    {
+                        bagState.PrepareDecryptBuffer(authSafeContents.Length);
+
+                        if (!bagState.LockedPassword)
+                        {
+                            bagState.LockPassword();
+                            int? workRemainingSave = workRemaining;
+
+                            try
+                            {
+                                contentData = DecryptSafeContents(
+                                    safeContentsAsn,
+                                    loaderLimits,
+                                    password,
+                                    ref bagState,
+                                    ref workRemaining);
+                            }
+                            catch (CryptographicException)
+                            {
+                                password = password.ContainsNull() ? "" : default;
+                                workRemaining = workRemainingSave;
+
+                                contentData = DecryptSafeContents(
+                                    safeContentsAsn,
+                                    loaderLimits,
+                                    password,
+                                    ref bagState,
+                                    ref workRemaining);
+                            }
+                        }
+                        else
+                        {
+                            contentData = DecryptSafeContents(
+                                safeContentsAsn,
+                                loaderLimits,
+                                password,
+                                ref bagState,
+                                ref workRemaining);
+                        }
+                    }
+                    else
+                    {
+                        // Should there be an option here to preserve this?
+                        // Ignore this?
```

Comment:
Since the design is now

* Count
* Extract parts
* Reassemble
* Send to OS

Unknown parts aren't a threat, so I don't think exploding is the right answer (despite being what it will do right now).  So the question is "do we add another Windows-specific 'leave some data in this PFX that we would otherwise remove' option?"  Long ago when I prototyped it, I was slightly thinking yes. But now I'm thinking ignore is probably the best bet (like we always did with the Unix PFX loader)

Reply:
DangerousNoLimits bypasses the extract and reassemble, correct? Does that mean that DNL is doing more than what its members expose? It will leave the data in the PFX and send it to the OS despite it not really having a property indicating that will happen.

Reply:
Correct.  Also, having thought about it more, I think that throwing is the right answer after all for unhandleable authsafe kinds; but perhaps with a "If you trust this file, you might be successful with DangerousNoLimits" type message.

Reply:
It does make me wonder if DNL will expose some edge cases since it will bypass unknown types. `new Pkcs12LoaderLimits(DangerousNoLimits).MacIterationLimit = 1000` will result in a loader limit that stops permitting unknown authsafe despite nothing changing the in the loader limit that would say otherwise.

Now, they are unknown authsafes that are probably inconsequential, but I'm using that as an example. Maybe there are other behaviors of PFXImportCertStore that we may not know about that DNL will allow but not really have an option indicating it as such.

I don't really know what I am proposing here, just musing out loud that it is an interesting quirk.

Reply:
Right now what I have is that DangerousNoLimits is a completely unique concept; bypassing the Extract/Reassemble path is triggered by reference equality to the singleton.

Getting less special would be doing value equality.

Getting slightly less special would be to add a property (strawman: `ProcessUnaltered`) that says that the OS will read the original data, not an Extract/Reassemble variant.  This solves a small set of problems, while creating others:

1) It's only true on Windows.  OpenSSL's p12_kiss is a little too "s" (pick one, either works); and macOS's loader is (was?) a few generations out of date and very picky, so I had to use extract/reassemble to make PFX work reliably on Macs.

2) We'd have to throw if ProcessUnaltered is true, but any of the Preserve values are false.  Or if ProcessUnaltered is true and IgnoreEncryptedAuthSafes is true.

3) "It makes this if weird".  While we know the OS will see it, we couldn't count it.  That also applies to unknown safebag types, where we might now be asking the OS to do uncounted work (though that's not currently true for any kind of safebag).  So, OK, we can also say that ProcessUnaltered:true isn't combinable with any of the numeric limits other than the MAC limit.  But, wow, this is getting weird.

Which leads me back to (A): just using reference equality, and saying "this concept is weird and special".

Reply:
Fair enough. If C# had discriminated unions, I would say it should look like:

```F#
type PkcsLoaderLimit =
    | DangerousNoLimits
    | Limited of macIterationLimit : int * maxKeys : int // etc
```

But, we don't, and they certainly don't work in .NET Framework even if they were to exist today. So, reference equality it is.